### ISSUE TITLE
After fixing an error and save buffer, lsp-ui still displays the error message

### ISSUE THREAD

**Issue by hisnawi** [emacs-lsp/lsp-ui#347]
I am enabling flycheck syntax checking only on save with :

>      flycheck-check-syntax-automatically  '(mode-enabled save)

And disabling lsp ui live reporting:

>           (setq lsp-ui-flycheck-live-reporting nil)

however, if I follow the steps:
edit a line and cause an error -> save buffer -> lsp-ui shows error message -> fix error -> save buffer -> lsp-ui still shows same error message 

If I check flycheck list of errors the error is not listed there anymore which tells me it is a lsp-ui issue.

**Comment by melton1968** [comment]
I was able to work around this issue with the following one line patch to `lsp-ui-flycheck.el`:

```
diff --git a/lsp-ui-flycheck.el b/lsp-ui-flycheck.el
index 3c48710..c5d5cb4 100644
--- a/lsp-ui-flycheck.el
+++ b/lsp-ui-flycheck.el
@@ -216,7 +216,6 @@ See https://github.com/emacs-lsp/lsp-mode."

 (defun lsp-ui-flycheck--report nil
   (and flycheck-mode
-       lsp-ui-flycheck-live-reporting
        (flycheck-buffer)))

 ;; FIXME: Provide a way to disable lsp-ui-flycheck
```

I am unfamiliar with the code-base, so I am uncertain if this is the correct fix, but it is working for me with the configuration listed above. Without this change, there is no way that flycheck can keep the errors messages in sync with the last result from the language server.

If you have installed lsp-ui from elpa, you can simply delete the line from the file and remove the corresponding .elc file and then restart emacs.

**Comment by hisnawi** [comment]
Hmm. it dosen't seem that worked for me.
I removed the line and reloaded emacs, but I still see live updates by flycheck.

**Comment by melton1968** [comment]
After a little more testing, I see that in general you are correct.  At this point, this is what I observe:

At global scope, I get live error reporting. 

In inner scopes (e.g. object abc { … }), I get the desired behavior.

So, the fix I suggested is clearly not correct, although it does makes the syntax checking much more usable for me since I am typing in some inner scope much more often that at global scope. I will continue with it until the real fix is in place.

ps. I presume you removed the .elc file?

**Comment by hisnawi** [comment]
There was no .elc file. Did you mean in the elpa/lsp-ui(version-number) directory?

**Comment by melton1968** [comment]
Yes, here is what I see in my emacs.d directory for lsp-ui  (MacOSX):

$ ls -al .emacs.d/elpa/lsp-ui-20191228.1151/
total 496
drwxr-xr-x  17 mmelton  staff    544 Dec 28 22:33 .
drwxr-xr-x  29 mmelton  staff    928 Dec 28 09:07 ..
-rw-r--r--   1 mmelton  staff   2220 Dec 28 09:06 lsp-ui-autoloads.el
-rw-r--r--   1 mmelton  staff  30760 Dec 28 09:06 lsp-ui-doc.el
-rw-r--r--   1 mmelton  staff  28389 Dec 28 09:06 lsp-ui-doc.elc
-rw-r--r--   1 mmelton  staff   1504 Dec 28 09:06 lsp-ui-doc.html
-rw-r--r--   1 mmelton  staff   9335 Dec 28 22:31 lsp-ui-flycheck.el
-rw-r--r--   1 mmelton  staff  12616 Dec 28 09:06 lsp-ui-imenu.el
-rw-r--r--   1 mmelton  staff  11405 Dec 28 09:06 lsp-ui-imenu.elc
-rw-r--r--   1 mmelton  staff  29922 Dec 28 09:06 lsp-ui-peek.el
-rw-r--r--   1 mmelton  staff  28452 Dec 28 09:06 lsp-ui-peek.elc
-rw-r--r--   1 mmelton  staff    482 Dec 28 09:06 lsp-ui-pkg.el
-rw-r--r--   1 mmelton  staff  24396 Dec 28 09:06 lsp-ui-sideline.el
-rw-r--r--   1 mmelton  staff  21438 Dec 28 09:06 lsp-ui-sideline.elc
-rw-r--r--   1 mmelton  staff   6105 Dec 28 09:06 lsp-ui.el
-rw-r--r--   1 mmelton  staff   5709 Dec 28 09:06 lsp-ui.elc

I manually removed the lsp-ui-flycheck.elc file and didn’t bother to re-compile it.

**Comment by hisnawi** [comment]
Not for me. No .elc files.

**Comment by melton1968** [comment]
After poking around the code a bit, here is my current understanding:

Diagnostic reports are received from the language server asynchronously and notifications can be received by using the `'lsp-after-diagnostics-hook`. We would like to update the `fly-check` errors only when the report from the language server was induced by saving the buffer. As far as I can tell, there is no direct way to determine this.

Trying to use `'after-save-hook` creates a race condition between `'after-save-hook` and `'lsp-after-diagnostics-hook`.

As an improved work-around for the time being, I made the following changes to `lsp-ui-flycheck.el` which uses the simple heuristic that if a diagnostics report arrives and the buffer is unmodified, then call `flycheck-buffer`.

Seems to work reasonably well.


```
diff --git a/lsp-ui-flycheck.el b/lsp-ui-flycheck.el
index 3c48710..eb6d91e 100644
--- a/lsp-ui-flycheck.el
+++ b/lsp-ui-flycheck.el
@@ -216,14 +216,14 @@ See https://github.com/emacs-lsp/lsp-mode."

 (defun lsp-ui-flycheck--report nil
   (and flycheck-mode
-       lsp-ui-flycheck-live-reporting
+       (or lsp-ui-flycheck-live-reporting
+	   (not (buffer-modified-p)))
        (flycheck-buffer)))

 ;; FIXME: Provide a way to disable lsp-ui-flycheck
 (defun lsp-ui-flycheck-enable (_)
   "Enable flycheck integration for the current buffer."
-  (when lsp-ui-flycheck-live-reporting
-    (setq-local flycheck-check-syntax-automatically nil))
+  (setq-local flycheck-check-syntax-automatically nil)
   (setq-local flycheck-checker 'lsp-ui)
   (lsp-ui-flycheck-add-mode major-mode)
   (add-to-list 'flycheck-checkers 'lsp-ui)
```

**Comment by yyoncho** [comment]
@melton1968 I was about to propose the same workaround. Would you like to create a PR?

>   (not (buffer-modified-p))

Here you should also check whether the flycheck mode is on saved because if it is manual the errors should be rendered only when the user calls `flycheck-buffer`.

**Comment by melton1968** [comment]
Yes, the current patch does not respect the value of flycheck-check-syntax-automatically and in fact sets it to nil to keep flycheck from checking the buffer before the diagnostics are received.

I can see how to support the following configurations with minimal code changes:

1. lsp-ui-flycheck-live-reporting is not nil
	lsp-ui-flycheck—report invokes (flycheck-buffer) every time new diagnostics are received

2. lsp-ui-flycheck-live-reporting is nil and flycheck-check-syntax-automatically contains ’save
	lsp-ui-flycheck—report  invokes (flycheck-buffer) when new diagnostics are received and the buffer is unmodified

3. lsp-ui-flycheck-live-reporting is nil and flycheck-check-syntax-automatically does not contain ’save
	lsp-ui-flycheck—report does not invoke (flycheck-buffer)


If that makes sense, I will make the changes and submit a PR.

**Comment by yyoncho** [comment]
> If that makes sense, I will make the changes and submit a PR.

It makes sense.

**Comment by hisnawi** [comment]
@melton1968 thank you for looking into this

**Comment by deviantfero** [comment]
Hi, I'm still seeing this issue, has this changed reached the melpa repository yet?

**Comment by melton1968** [comment]
I updated my packages from ELPA stable and that version had the fix.

cheers,
mark

**Comment by hisnawi** [comment]
Melton, could you share the package version number?

**Comment by melton1968** [comment]
lsp-ui             20200106.1152

cheers,
mark

**Comment by hisnawi** [comment]
Interesting...now flycheck doesn't work at all. Whether I save the file or I don't.
It gives the no error symbol no matter what I code.
This is my setup:

```
  (use-package lsp-ui
    :ensure t
    :after lsp-mode
    :diminish
    :commands lsp-ui-mode
    :bind (("s-c" . lsp-ui-imenu))
    :custom-face
    (lsp-ui-doc-background ((t (:background nil))))
    (lsp-ui-doc-header ((t (:inherit (font-lock-string-face italic)))))
    :custom
    (lsp-ui-doc-enable nil)   ;; Disable since it is distracting (all following apply when enabled)
    (lsp-ui-doc-include-signature t)
    (lsp-ui-doc-position 'top)
    (lsp-ui-doc-border (face-foreground 'default))
    ;;(lsp-ui-sideline-enable nil)
    (lsp-ui-sideline-ignore-duplicate t)
    (lsp-ui-sideline-show-code-actions nil)
    :config
    ;; Use lsp-ui-doc-webkit only in GUI
    (setq lsp-ui-doc-use-webkit t)
    ;; WORKAROUND Hide mode-line of the lsp-ui-imenu buffer
    ;; https://github.com/emacs-lsp/lsp-ui/issues/243
    (defadvice lsp-ui-imenu (after hide-lsp-ui-imenu-mode-line activate)
      (setq mode-line-format nil))
    )
  (use-package flycheck
    :ensure t
    :init
    (add-hook 'prog-mode-hook #'flycheck-mode)
    :config
    (setq-default
     flycheck-display-errors-delay        0.3
     flycheck-check-syntax-automatically  '(mode-enabled save)
     flycheck-disabled-checkers           '(emacs-lisp-checkdoc c/c++-clang c/c++-cppcheck c/c++-gcc)
     )
    (setq-default flycheck-gcc-language-standard "c++17")
    (setq-default flycheck-clang-language-standard "c++17")
    )

```

**Comment by melton1968** [comment]
I just loaded up a c++ file to verify and I get syntax checking. Here is the relevant part of my .emacs

;; Enable nice rendering of diagnostics like compile errors.
(use-package flycheck
  :init (global-flycheck-mode)
  :config
  (setq flycheck-check-syntax-automatically '(save))
  (setq flycheck-clang-args (list "-extra-arg-before=-xc++")))

(use-package lsp-mode
  ;; Optional - enable lsp-mode automatically in scala files
  :hook ((c++-mode . lsp) (scala-mode . lsp))
  :config
  (setq lsp-enable-snippet nil)
  (setq lsp-prefer-flymake nil)
  (setq lsp-clients-clangd-args (list "--query-driver=/opt/local/bin/clang*9.0"))
  (setq lsp-clients-clangd-executable "/opt/local/bin/clangd-mp-9.0"))

(use-package lsp-ui
  :config
  (setq lsp-ui-flycheck-live-reporting nil)
  )

;; Add snippets and company-lsp backend for metals
(use-package yasnippet)
(use-package company-lsp)

cheers,
mark

**Comment by melton1968** [comment]
And, here is what I get when I run flycheck-verify-setup:

Syntax checkers for buffer rest_t0.cpp in c++-mode:

First checker to run:

  lsp-ui (explicitly selected)
    - may enable: yes
    - predicate:  t

Checkers that could run if selected:

  c/c++-clang  select
    - may enable:    yes
    - executable:    Found at /opt/local/bin/clang
    - next checkers: c/c++-cppcheck

  c/c++-gcc  select
    - may enable:    yes
    - executable:    Found at /usr/bin/gcc
    - next checkers: c/c++-cppcheck

Checkers that are compatible with this mode, but will not run until properly configured:

  c/c++-cppcheck (disabled)
    - may enable: Automatically disabled!
    - executable: Not found

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 32snapshot (package: 20191126.2142)
Emacs version:    26.3
System:           x86_64-apple-darwin19.0.0
Window system:    nil

cheers,
mark
